bindings/matrix-sdk-ffi/src/client.rs

```diff
@@ -131,31 +155,56 @@ impl Client {
             let sync_settings = SyncSettings::new().filter(Filter::FilterId(&filter_id));
 
             client
-                .sync_with_callback(sync_settings, |sync_response| async {
-                    if !state.read().unwrap().has_first_synced {
-                        state.write().unwrap().has_first_synced = true;
-                    }
-
-                    if state.read().unwrap().should_stop_syncing {
-                        state.write().unwrap().is_syncing = false;
-                        return LoopCtrl::Break;
-                    } else if !state.read().unwrap().is_syncing {
-                        state.write().unwrap().is_syncing = true;
-                    }
-
-                    if let Some(delegate) = &*delegate.read().unwrap() {
-                        delegate.did_receive_sync_update()
-                    }
-
-                    if let Some(session_verification_controller) =
-                        &*session_verification_controller.read().await
-                    {
-                        session_verification_controller
-                            .process_to_device_messages(sync_response.to_device)
-                            .await;
+                .sync_with_callback(sync_settings, |result| async {
+                    if let Ok(sync_response) = result {
+                        if !state.read().unwrap().has_first_synced {
+                            state.write().unwrap().has_first_synced = true;
+                        }
+
+                        if state.read().unwrap().should_stop_syncing {
+                            state.write().unwrap().is_syncing = false;
+                            return LoopCtrl::Break;
+                        } else if !state.read().unwrap().is_syncing {
+                            state.write().unwrap().is_syncing = true;
+                        }
+
+                        if let Some(delegate) = &*delegate.read().unwrap() {
+                            delegate.did_receive_sync_update()
+                        }
+
+                        if let Some(session_verification_controller) =
+                            &*session_verification_controller.read().await
+                        {
+                            session_verification_controller
+                                .process_to_device_messages(sync_response.to_device)
+                                .await;
+                        }
+
+                        LoopCtrl::Continue
+                    } else {
```

Comment:
Should probably break this off into its own method

Reply:
Updated.